### PR TITLE
dont graph break on list[SymInt] comparisons

### PR DESCRIPTION
**Pull request by bdhirsh**
Make sure that we don't graph break on code like:
```
assert a.size() == b.size()
```

When a and b have dynamic shapes.

Two things to call out:

(1) The SymBool PR hasn't landed yet, so I just added `operator.__and__` support to SymInt as a dumb workaround. We can either remove that when SymBool lands, or wait for SymBool to land before merging this

(1) There was some fiddling I had to do for ListVariable, that I didn't have to do for e.g. SizeVariable. Comparing two lists is supposed to return a bool, but ListVariable is represented as a List of proxies - so I had to manually reduce across the list.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92643
* #92618
* __->__ #92617

### CODE REVIEW

File: torch/__init__.py
Comment (bdhirsh) [inline torch/__init__.py:266]: open to either waiting for the SymBool PR to merge first, or landing this and removing `__and__` from SymInt later.

**Comment by ezyang** [comment]
I'm nearly ready to land symbool modulo review so let's wait

File: torch/_dynamo/symbolic_convert.py
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:964]: I think we want to do this recursively. If we have two lists, AND we've specialized on their sizes (I think this is always true for ListVariable?) we want to recursively run comparison on each of their elements. What we essentially have done is reimplemented the `__eq__` implementation of the list itself, because that's ostensibly in CPython C itself (specifically, `PyObject_RichCompare`), however, the *inside* should be delegated to whatever other knowledge we have.

File: torch/_dynamo/symbolic_convert.py
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:958]: This is wrong, you'll report true if you compare a list and a tuple but these are never equal 

```
>>> [1] == (1,)
False
```

File: torch/_dynamo/symbolic_convert.py
Comment (ezyang) [reply torch/_dynamo/symbolic_convert.py:964]: This does kind of suggest we need some sort of "tp_richcompare" concept on VariableTracker, maybe, to keep track of this sort of thing. cc if @jansel has any thoughts

Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:964]: I think you support too many ops here. If I do an inequality it needs to be treated lexicographically. But I don't see any such logic. You probably didn't care about support inequalities, but they are technically supported here.

File: torch/_dynamo/symbolic_convert.py
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:999]: Creating a DynamicShapeVariable here is kind of pointless, because the comparison ops MUST return a bool, which means you have guarded on the result of the comparison here. I think guarding is the right thing to do here, and you should just resolve this down to a constant at this point (with the understanding that the guards will help you out here later)

File: torch/__init__.py
Comment (bdhirsh) [reply torch/__init__.py:266]: oof good point

File: torch/_dynamo/symbolic_convert.py
Comment (bdhirsh) [reply torch/_dynamo/symbolic_convert.py:964]: Doing it recursively sounds good to me. Although speaking of this comment:

> This is wrong, you'll report true if you compare a list and a tuple but these are never equal

Obviously I could include a check that they're the same type here, but I agree there are an annoying amount of checks here. Is there a way that you can think of to avoid having dynamo add a manual check like that?

Comment (bdhirsh) [reply torch/_dynamo/symbolic_convert.py:964]: Good point. I guess I'll just reduce the number of ops I support here. Although I'm still curious, similar to the above comment - is there a way to get the different types of list equalities to be handled more automagically? Or are we stuck re-implementing stuff like that in dynamo.

File: torch/_dynamo/symbolic_convert.py
Comment (bdhirsh) [reply torch/_dynamo/symbolic_convert.py:999]: So my thought was - once SymBool lands, then it'll be okay for the comparison op to return a SymBool, thus requiring our tracker to be a DynamicShapeVariable. Is that wrong? (comparison ops on SymInts should unconditionally materialize the concrete bool).

I guess in practice, almost every comparison on SymInts will do some kind of branching on the bool afterwards

File: torch/_dynamo/symbolic_convert.py
Comment (ezyang) [reply torch/_dynamo/symbolic_convert.py:964]: The "proper" way to do this is to replicate the logic that happens in CPython (which I am not entirely sure how it works). Something quick like only handling if the type() is the same, that seems fine (because it's always sound to unsupported otherwise)

Comment (ezyang) [reply torch/_dynamo/symbolic_convert.py:964]: Anything that is morally implemented in CPython C code must be reimplemented. That being said, we can always create local abstractions to reduce duplication, etc.

File: torch/_dynamo/symbolic_convert.py
Comment (ezyang) [reply torch/_dynamo/symbolic_convert.py:999]: Yeah, I think in principle it is better to support symbolic bools in dynamo and propagate as far as they can.

File: torch/fx/experimental/symbolic_shapes.py
Comment (jansel) [inline torch/fx/experimental/symbolic_shapes.py:286]: is `__or__` also like this? Might be good to do a quick scan through the operator docs.

File: torch/_dynamo/symbolic_convert.py
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:960]: suggest a helper method to handle this repeated "all constant or dynamic shape" check

File: torch/_dynamo/symbolic_convert.py
Comment (jansel) [reply torch/_dynamo/symbolic_convert.py:964]: I think we could just add a `compare()` method to `VariableTracker` that raises an exception for unhandled.  I think a lot of the complexity would be reduced if we used method dispatch instead of `isinstance`.  It would also make the recursive check way easier.  For example, BaseListVariable could just check type and length then recursively call compare on all the items.

File: torch/_dynamo/symbolic_convert.py
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:985]: Could we reuse some existing code by doing `BuiltinVariable(all).call_function(....)`

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/symbolic_convert.py
Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:964]: +1 this should be on VariableTracker

**Comment by bdhirsh** [comment]
will do (not quite ready yet, pushed my local changes mostly so Joel could take a look)

File: torch/_dynamo/variables/lists.py
Comment (bdhirsh) [inline torch/_dynamo/variables/lists.py:108]: The way I implemented this was by writing a generic list comparison (I think tuple vs. list comparison ops all do the same type of reduction?). The bits that I made different are:
- Each VariableTracker gets its own type check (tuple must be compared with tuple, list with list, but torch.Size vs. tuple comparison is ok)
- I implemented `curr, rest = next_variables()` on a few BaseListVariableTracker subclasses to do the reduction, which is container-dependent.

File: torch/_dynamo/variables/lists.py
Comment (bdhirsh) [inline torch/_dynamo/variables/lists.py:123]: Added a separate base case for 0 vs 1, so I could avoid adding handling for `operator.and_(DynamicShapeVariable, ConstantVariable)` (figured it was annoying to special case, and also is rare in user code)

**Comment by bdhirsh** [comment]
This is ready for another round (but not the other PR's in the stack)

File: torch/_dynamo/variables/functions.py
Comment (ezyang) [inline torch/_dynamo/variables/functions.py:270]: This is just... Python constant isn't it

File: torch/_dynamo/variables/tensor.py
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:482]: Don't you have to swap the op?

File: torch/_dynamo/variables/lists.py
Comment (ezyang) [inline torch/_dynamo/variables/lists.py:121]: ... why this part recursive lol

File: test/dynamo/test_misc.py
Comment (voznesenskym) [inline test/dynamo/test_misc.py:153]: this test does not invoke generic_list_compare?

Comment (voznesenskym) [reply test/dynamo/test_misc.py:153]: Am wrong, does in dynamic shapes

File: torch/_dynamo/variables/lists.py
Comment (voznesenskym) [reply torch/_dynamo/variables/lists.py:105]: Changed impl

File: torch/_dynamo/variables/lists.py
Comment (bdhirsh) [inline torch/_dynamo/variables/lists.py:133]: Curiosity question: when does checkpointing happen? (Just thinking about it, I think my guess would be it's at the end of processing each python byte code, so that a single call to COMPARE_OP is "all-or-nothing" success (write every torch op to the graph) or failure (graph break).

File: torch/_dynamo/variables/lists.py
Comment (bdhirsh) [inline torch/_dynamo/variables/lists.py:140]: I wonder if this'll bite us in the near future, in a world where users mark certain dimensions as dynamic explicitly (so it could be common to compare two torch.Size arguments, that have a mix of `ConstantVariable` and `DynamicVariable`.

If you think it's worth not taking care of for now, that sounds fine to me.

File: torch/_dynamo/variables/lists.py
Comment (bdhirsh) [inline torch/_dynamo/variables/lists.py:163]: I found this to be a slightly weird convention, that `wrap_fx_proxy` is a "blessed" boilerplate-free way to create a `TensorVariableTracker`, while every other tracker needs the extra boilerplate. But I think I originally had this in the PR too 😛. Doesn't need to change in this PR.

File: torch/_dynamo/variables/lists.py
Comment (voznesenskym) [reply torch/_dynamo/variables/lists.py:140]: I think it's probably okay for now? I can relax this assertion more, let me do that - I was just trying to exclude tensors.

File: torch/_dynamo/variables/lists.py
Comment (voznesenskym) [reply torch/_dynamo/variables/lists.py:133]: Grep for the checkpoint/Checkpointable structure :) 

`copy_graphstate` creates new ones

File: torch/_dynamo/variables/lists.py
Comment (voznesenskym) [reply torch/_dynamo/variables/lists.py:159]: btw wrap_fx_proxy's blessed sibling for non Tensors is wrap_fx_proxy_cls - its usually for more complex types, but I agree, we should standardize.

The history here is this used to be TensorVariable's .create method!

File: torch/_dynamo/symbolic_convert.py
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:967]: This `or` seems a bit inconsistent.  If `left` fails the check we will call the base variable tracker `compare()` (that just raises the same error as the `else` down below).

Given we define a base class version of compare, we could just call this unconditionally.

File: torch/_dynamo/symbolic_convert.py
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:977]: On second though, I'd suggest trying to reuse some existing machinery for this rather than adding a new `compare()` method.

We already have:

1)
```
left.call_method(tx, "__eq__", [right], {})
```

2)
```
BuiltinVariable(operator.eq).call_function(tx, [left, right], {})
```

Often option 2 will redispatch to option 1:
https://github.com/pytorch/pytorch/blob/0f802eedc2671fd17106923ef2981e3e6eaa8b95/torch/_dynamo/variables/builtin.py#L547-L563

This has the advantage that it will work if the user calls `__eq__` or `operator.eq` directly.

We should be able to just make ths:
```suggestion
        else:
            self.push(BuiltinVariable(supported_any[op], **options).call_function(tx, [left, right], {}))
```

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/symbolic_convert.py
Comment (ezyang) [reply torch/_dynamo/symbolic_convert.py:977]: This doesn't change the suggestion, but @jansel I want to draw your attention to an ongoing conversation we're having at https://github.com/pytorch/pytorch/pull/93271 The short version:

1. For full fidelity operator.add handling, we need "reversibility" logic (that's what #93271 is about) to try both `__add__` and then try `__radd__` if add is NotImplemented
2. For FX graph tracing purposes, we prefer to trace into the graph at the operator.add level, not the `__add__` method call level, for a variety of reasons, but a very pedestrian reason is when you do reversibility logic it looks really ugly to add FX nodes for both `__add__` and `__radd__`

cc @jbschlosser I wonder if `BINARY_ADD` and similar opcodes go via the builtin variables

Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:977]: Sure. We can do it through call_eq, call_gt, etc. However, you still need to have a per-object expansion of equality comparison - I am not sure there is a way of getting around .compare() calls or __eq__ calls or equivalent on the VariableTrackers themselves. 

So even if we rewrote this to be BuiltinVariable's roughly `call_{supported_any[op]}` - it would not preclude the need to compare in [left] [right] - in our case, which would still need to be delegated to ListVariable's `__eq__` or `compare` (same thing), which in turn would need to call that again on the contained (`.items`) objects. 

Would you agree with that?

Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:977]: > This doesn't change the suggestion, but @jansel I want to draw your attention to an ongoing conversation we're having at #93271 The short version:
> 
> 1. For full fidelity operator.add handling, we need "reversibility" logic (that's what [General reversible binary op support (e.g. __add__ / __radd__) in dynamo #93271](https://github.com/pytorch/pytorch/pull/93271) is about) to try both `__add__` and then try `__radd__` if add is NotImplemented
> 2. For FX graph tracing purposes, we prefer to trace into the graph at the operator.add level, not the `__add__` method call level, for a variety of reasons, but a very pedestrian reason is when you do reversibility logic it looks really ugly to add FX nodes for both `__add__` and `__radd__`
> 
> cc @jbschlosser I wonder if `BINARY_ADD` and similar opcodes go via the builtin variables

I said this in chat - but if we do this, we might as well go all the way and break up the giant if/else trees in `call_method` and `call_function` with delegation to ops-defined-on-subclass-variable-trackers like this...

Comment (jansel) [reply torch/_dynamo/symbolic_convert.py:977]: @voznesenskym I think a helper method makes sense, but it should be local to list variable (not a new API on every single Variable Tracker)

@ezyang I'll make some comments over there (once I finish going through).  I wasn't trying to make a strong argument for function vs method in this case, more of a "don't and a third option"

Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:977]: > @voznesenskym I think a helper method makes sense, but it should be local to list variable (not a new API on every single Variable Tracker)

Agreed, the `compare` barely do anything anyway, just grab the args and essentially make a new proxy for the comparison op.

Comment (jbschlosser) [reply torch/_dynamo/symbolic_convert.py:977]: > cc @jbschlosser I wonder if BINARY_ADD and similar opcodes go via the builtin variables

They do:
https://github.com/pytorch/pytorch/blob/37f7c00a8aae2ced9666a69d0d51a4d9d4b379af/torch/_dynamo/symbolic_convert.py#L1462

https://github.com/pytorch/pytorch/blob/37f7c00a8aae2ced9666a69d0d51a4d9d4b379af/torch/_dynamo/symbolic_convert.py#L133-L141

File: torch/_dynamo/variables/lists.py
Comment (jansel) [inline torch/_dynamo/variables/lists.py:130]:
```suggestion
        for l, r in zip(left.items, right.items):
```

File: torch/_dynamo/variables/lists.py
Comment (jansel) [inline torch/_dynamo/variables/lists.py:145]: I believe the entire rest of this function can just be replaced with:
```suggestion
        return functools.reduce(
            lambda a, b: BuiltinVariable(operator.and_).call_function(tx, [a, b], {}),
            comps
        )
```

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/lists.py
Comment (jansel) [inline torch/_dynamo/variables/lists.py:132]: With the change below, this might actually be working now.

**Review by jansel — APPROVED** [review]

**Comment by voznesenskym** [comment]
Mild regret accepting ci comments, I think it messes with ghstack, had to redo a few of them cause of it. Should be fine now...

**Comment by voznesenskym** [comment]
@malfet could CI be importing the wrong sha? or having a ghstack hiccup? the line numbers locally are not lining up ...

**Comment by ezyang** [comment]
You cannot accept suggestions in GitHub with ghstack, please do not do that

**Comment by ezyang** [comment]
To check if the trees are consistent, look at gh/bdhirsh/367/head versus gh/bdhirsh/367/orig tree hash.

```
(/home/ezyang/local/a/pytorch-env) [ezyang@devgpu020.ftw1 ~/local/a/pytorch (a6a482e0)]$ git rev-parse origin/gh/bdhirsh/367/head:
beccb2ebac4c447ad10f8063de7771de3e5b65cb
(/home/ezyang/local/a/pytorch-env) [ezyang@devgpu020.ftw1 ~/local/a/pytorch (a6a482e0)]$ git rev-parse origin/gh/bdhirsh/367/orig:
beccb2ebac4c447ad10f8063de7771de3e5b65cb
```

they are consistent so it is not a ghstack problem

**Comment by voznesenskym** [comment]
> To check if the trees are consistent, look at gh/bdhirsh/367/head versus gh/bdhirsh/367/orig tree hash.
> 
> ```
> (/home/ezyang/local/a/pytorch-env) [ezyang@devgpu020.ftw1 ~/local/a/pytorch (a6a482e0)]$ git rev-parse origin/gh/bdhirsh/367/head:
> beccb2ebac4c447ad10f8063de7771de3e5b65cb
> (/home/ezyang/local/a/pytorch-env) [ezyang@devgpu020.ftw1 ~/local/a/pytorch (a6a482e0)]$ git rev-parse origin/gh/bdhirsh/367/orig:
> beccb2ebac4c447ad10f8063de7771de3e5b65cb
> ```

> they are consistent so it is not a ghstack problem

Well, something happened that cause local, github, and CI to deviate, so I made a new PR.

**Comment by voznesenskym** [comment]
> @pytorchbot merge -f "Failure: There is only 5431220KB free space left in /, which is less than the minimum requirement of 6291456KB for macOS"

Actually nvm, I dont want this, canceled

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/pull/94054

**Pull request closed without merging**
